src/mlpack/core/tree/binary_space_tree/rp_tree_mean_split_impl.hpp

```diff
@@ -0,0 +1,167 @@
+/**
+ * @file rp_tree_mean_split_impl.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Implementation of class (RPTreeMeanSplit) to split a binary space partition
+ * tree.
+ */
+#ifndef MLPACK_CORE_TREE_BINARY_SPACE_TREE_RP_TREE_MEAN_SPLIT_IMPL_HPP
+#define MLPACK_CORE_TREE_BINARY_SPACE_TREE_RP_TREE_MEAN_SPLIT_IMPL_HPP
+
+#include "rp_tree_max_split.hpp"
+
+namespace mlpack {
+namespace tree {
+
+template<typename BoundType, typename MatType>
+bool RPTreeMeanSplit<BoundType, MatType>::SplitNode(const BoundType&  bound,
+                                                  MatType& data,
+                                                  const size_t begin,
+                                                  const size_t count,
+                                                  SplitInfo& splitInfo)
+{
+  const size_t maxNumSamples = 100;
+  const size_t numSamples = std::min(maxNumSamples, count);
+  arma::uvec samples;
+
+  // Get no more than numSamples distinct samples.
+  math::ObtainDistinctSamples(begin, begin + count, numSamples, samples);
+
+  // Find the average distance between points.
+  ElemType averageDistanceSq = GetAveragePointDistance(data, samples);
```

Comment:
Continuing with the discussion about whether the theory is still valid with your modifications, I think that you are right to do sampling to estimate statistics of the dataset.  An exact approach would be quadratic time!  This would yield a total tree building time of something like O(N^2 log N).  So I think that implementing the tree exactly as given in the paper is not really feasible.

It should be possible to re-derive the theorems, taking into account the sampling you are doing here, without affecting the results significantly (only the success probability).  Don't feel obligated to do that---I am mostly just thinking out loud here about the implications. :)